qualang_tools/config/helper_tools.py

```diff
@@ -0,0 +1,390 @@
+        :param operation_name: name of the operation to get the waveforms from. Must be defined in the config.
+        :param new_name: name of the new operation.
+        """
+        # Check inputs
+        if element not in self.data["elements"].keys():
+            raise KeyError(f"The element '{element}' in not defined in the config.")
+        if operation_name not in self.data["elements"][element]["operations"].keys():
+            raise KeyError(
+                f"The operation '{operation_name}' in not defined in the config for the element '{element}'."
+            )
+        # Copy operation
+        pulse_name = self.data["elements"][element]["operations"][operation_name]
+        self.data["pulses"][element + "_" + new_name + self.pulse_suffix] = _deepcopy(
+            self.data["pulses"][pulse_name]
+        )
+        # Rename the copied operation
+        self.data["elements"][element]["operations"][new_name] = (
+            element + "_" + new_name + self.pulse_suffix
+        )
+
+    def update_waveforms(
+        self,
+        element: str,
+        operation_name: str,
+        wf: Tuple[List[float], List[float]] or Tuple[List[float]],
+    ):
+        """
+        Update the waveforms from a specific operation and element.
+
+        :param element: name of the element to get the waveforms from. Must be defined in the config.
+        :param operation_name: name of the operation to get the waveforms from. Must be defined in the config.
+        :param wf: Tuple containing the updated waveforms. (wf_single,) for a singleInput element and (wf_i, wf_q) for a mixInputs element.
+        """
+        # Check inputs
+        if element not in self.data["elements"].keys():
+            raise KeyError(f"The element '{element}' in not defined in the config.")
+        if operation_name not in self.data["elements"][element]["operations"].keys():
+            raise KeyError(
+                f"The operation '{operation_name}' in not defined in the config for the element '{element}'."
+            )
+        if "mixInputs" not in self.data["elements"][element].keys() and len(wf) == 2:
+            raise KeyError(f"Element '{element}' is not a mixInputs element.")
+        if "singleInput" not in self.data["elements"][element].keys() and len(wf) == 1:
+            raise KeyError(f"Element '{element}' is not a singleInput element.")
+        if len(wf) == 2:
+            if len(wf[0]) != len(wf[1]):
+                raise ValueError("The 'I' and 'Q' waveforms must have the same length.")
+
+        pulse_name = self.data["elements"][element]["operations"][operation_name]
+        self.data["pulses"][pulse_name]["length"] = len(wf[0])
+        # mixInputs element
+        if len(wf) == 2:
+            wf_name = element + "_" + operation_name + self.wf_suffix
+            wf_i = wf[0]
+            wf_q = wf[1]
+            if len(wf_i) > 0:
+                if wf_i[:-1] == wf_i[1:]:
```

Comment:
I think that you are checking if the waveforms are symmetric, not if they are constant. For example, a gaussian or a cosine wave would be detected here as "constant".
This has some nice solutions for checking if all elements are equal:
https://stackoverflow.com/questions/3844801/check-if-all-elements-in-a-list-are-identical

Reply:
Not really, it compares two lists shifted by one item so if all elements are equal, then it will be True and if the list is symmetric then it will be False because all neighbors won't be equal. You can test it with `wf = [0, 1, 0]` where it compares [0,1] with [1,0]

Reply:
My bad